serverless_sim/src/metric.rs

```diff
@@ -55,34 +58,72 @@ impl MechMetric {
             node_task_new_cnt: HashMap::new(),
         }
     }
+
+    // FIX 滑动窗口有问题，需要重新设计
+    // 新请求生成之后
     pub fn on_new_req_generated(&mut self, env: &SimEnv) {
+
+        // 清空数据表
         self.fn_unsche_req_cnt.clear();
-        self.fn_recent_req_cnt_window.clear();
         self.node_task_new_cnt.clear();
 
+        // 记录这一帧中每个函数的未被调度的请求数量
+        let mut fn_count = HashMap::new();
+
+        // 遍历每个请求
         for (_, req) in env.core.requests().iter() {
+            // DAG的迭代器
             let mut walker = env.dag(req.dag_i).new_dag_walker();
+
+            // 遍历DAG的每个节点
             while let Some(fngid) = walker.next(&env.dag(req.dag_i).dag_inner) {
+
+                // 获取函数节点id
                 let fnid = env.dag_inner(req.dag_i)[fngid];
+
+                // 若该函数没有被调度到节点上
                 if req.get_fn_node(fnid) == None {
+
+                    // 如果fnid在fn_unsche_req_cnt中存在，则将其值加一，否则插入一个新值1
                     self.fn_unsche_req_cnt
                         .entry(fnid)
                         .and_modify(|v| *v += 1)
                         .or_insert(1);
                 }
+
+                // 如果请求的已完成函数列表中不包含fnid，则计数+1
                 if !req.done_fns.contains_key(&fnid) {
```

Comment:
You can make the window step into two:
- Collect each fn's **req_cnt** in a map
- For each fn, if fn is not in map, push zero to window; if fn is in map, push the **req_cnt** to window